routing/payment_lifecycle.go

```diff
@@ -26,189 +28,524 @@ func (e errNoRoute) Error() string {
-	paySession     PaymentSession
-	timeoutChan    <-chan time.Time
-	currentHeight  int32
-	finalCLTVDelta uint16
-	attempt        *channeldb.HTLCAttemptInfo
-	circuit        *sphinx.Circuit
-	lastError      error
+	router        *ChannelRouter
+	totalAmount   lnwire.MilliSatoshi
+	paymentHash   lntypes.Hash
+	paySession    PaymentSession
+	timeoutChan   <-chan time.Time
+	currentHeight int32
 }
 
 // resumePayment resumes the paymentLifecycle from the current state.
 func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
+	shardHandler := &shardHandler{
+		router:      p.router,
+		paymentHash: p.paymentHash,
+		quit:        make(chan struct{}),
+	}
+
+	// When the payment lifecycle loop exits, we make sure to signal any
+	// sub goroutine to exit, then wait for them to return.
+	defer func() {
+		close(shardHandler.quit)
+		shardHandler.wg.Wait()
+	}()
+
+	// Each time we send a new payment shard, we'll spin up a goroutine
+	// that will collect the result. Either a payment result will be
+	// returned, or a critical error signaling that we should immediately
+	// exit.
+	shardOutcomes := make(chan *shardOutcome)
+	criticalErrs := make(chan error)
+	var lastError error
+
+	// If we had any existing attempts outstanding, we'll start by and
+	// resume spinning up goroutines that'll collect their results and
+	// deliver them to the lifecycle loop below.
+	existingAttempts, err := p.router.cfg.Control.GetAttempts(
+		p.paymentHash,
+	)
+	if err != nil {
+		return [32]byte{}, nil, err
+	}
+
+	for _, a := range existingAttempts {
+		attempt := a
+
+		shardHandler.wg.Add(1)
+		go func() {
+			defer shardHandler.wg.Done()
+
+			outcome, err := shardHandler.collectResult(
+				&attempt.HTLCAttemptInfo,
+			)
+			if err != nil {
+				select {
+				case criticalErrs <- err:
+				case <-shardHandler.quit:
+				}
+				return
+			}
+
+			select {
+			case shardOutcomes <- outcome:
+			case <-shardHandler.quit:
+			}
+		}()
+	}
+
+	var (
+		success         = false
+		routeFailure    *paymentFailure
+		terminalFailure *paymentFailure
+	)
+
 	// We'll continue until either our payment succeeds, or we encounter a
 	// critical error during path finding.
+Loop:
 	for {
 
-		// If this payment had no existing payment attempt, we create
-		// and send one now.
-		if p.attempt == nil {
-			firstHop, htlcAdd, err := p.createNewPaymentAttempt()
-			if err != nil {
-				return [32]byte{}, nil, err
+		// Check if we got any terminal failure.
+		var failure *paymentFailure
+		switch {
+		case terminalFailure != nil:
+			failure = terminalFailure
+
+		// We only consider the route failure if we didn't have another
+		// terminal failure.
+		case routeFailure != nil:
+			failure = routeFailure
+		}
+
+		// Get all attempts made for this payment, and extract the ones
+		// still in flight.
+		attempts, err := p.router.cfg.Control.GetAttempts(
+			p.paymentHash,
+		)
+		if err != nil {
+			return [32]byte{}, nil, err
+		}
+
+		var activeShards int
```

Comment:
Actual number seems not to be used, can be a bool?

Reply:
done,